### ISSUE TITLE
A problem in gardgets `splite_base`

### ISSUE THREAD

**Issue by xiyu1984** [0xPolygonZero/plonky2#1421]
If we use [`le_sum`](https://github.com/0xPolygonZero/plonky2/blob/bfcfcdb498d253ff33121dba03a130be89fcc96c/plonky2/src/gadgets/split_base.rs#L42) to convert a bits target, that is a `Vec<BoolTarget>`, to a field target (Target related to a `GoldilocksField`), some values of the GoldilocksField will be invalid. 

That is, we cannot set the target with the values in `[2^63, 0xFFFFFFFF00000000]`

**Comment by xiyu1984** [comment]
If we change the code: 
```rust
assert!(
            num_bits <= log_floor(F::ORDER, 2),
            "{} bits may overflow the field",
            num_bits
);
```
to
```rust
assert!(
            num_bits <= log2_ceil(F::ORDER as usize),
            "{} bits may overflow the field",
            num_bits
);
```

And create a `Target` related to a `GoldilocksField` number through [le_sum](https://github.com/0xPolygonZero/plonky2/blob/bfcfcdb498d253ff33121dba03a130be89fcc96c/plonky2/src/gadgets/split_base.rs#L38).

And set the target in the witness to any number larger than `2^63` and less than `F::ORDER`, the `prove` will work but the `verify` will fail. The error is as follows:

![1702365565065](https://github.com/0xPolygonZero/plonky2/assets/83746881/4e5eb047-647b-43c2-a73c-a734f404c830)

**Comment by wborgeaud** [comment]
You are right that Goldilocks field elements with 64 bits are not reachable with this gadget, but we want to make sure it's impossible to overflow and this is only possible by rejecting 64 bit vectors altogether.
In Plonky2 we only use this gadget with smaller vectors, so this isn't an issue for us. Do you have a use case where supporting 64 bit vectors is necessary?

**Comment by xiyu1984** [comment]
> You are right that Goldilocks field elements with 64 bits are not reachable with this gadget, but we want to make sure it's impossible to overflow and this is only possible by rejecting 64 bit vectors altogether. In Plonky2 we only use this gadget with smaller vectors, so this isn't an issue for us. Do you have a use case where supporting 64 bit vectors is necessary?

Yeah, in my case I need to prove the calculation of `u128` numbers, and I found the [BigUintTarget](https://github.com/0xPolygonZero/plonky2-ecdsa/blob/bdb6504bca250db1548cdcce2407d4e334990c33/src/gadgets/biguint.rs#L22) can do this. But in order to created the `Poseidon` hash of some data including the `u128` numbers, I need to convert `BigUintTarget` to two `Target` related to two GoldilocksField values. 

In order to do the converting, I transfer the `BigUintTarget` into bits first, and then use `le_sum` to generate the `u128 target`. Sometimes the 64 bits `BoolTarget` vec reaches.

Anyway, luckily, I have solved this problem today, without any changed to the `plonky2 repo`. 

The compromised solution is that, I generate a target with the first 63 bits, and add it to a constant target with the value `` << 63` if the highest bit is 1.

**Comment by wborgeaud** [comment]
Ok great, will close this then.
